packages/material-ui-styles/src/withStyles.js

```diff
@@ -0,0 +1,339 @@
+        this.props.stylesOptions.sheetsManager,
+        stylesCreator,
+        this.props.theme,
+      );
+      if (sheetManager.dynamicSheet) {
+        console.log('sheetManager', sheetManager);
+        sheetManager.dynamicSheet.update(this.props);
+      }
+    }
+
+    componentWillUnmount() {
+      detach({
+        theme: this.props.theme,
+        stylesOptions: this.props.stylesOptions,
+        stylesCreator,
+      });
+    }
+
+    render() {
+      const { classes, theme, innerRef, stylesOptions, ...other } = this.props;
+
+      const oldTheme = this.theme;
+      this.theme = theme;
+
+      if (oldTheme !== theme) {
+        attach({
+          props: this.props,
+          theme,
+          stylesOptions,
+          stylesCreator,
+          name,
+        });
+
+        if (oldTheme) {
+          // Rerender the component so the underlying component gets the theme update.
+          // By theme update we mean receiving and applying the new class names.
+          setTimeout(() => {
+            detach({
+              theme: oldTheme,
+              stylesOptions,
+              stylesCreator,
+            });
+          });
+        }
+      }
+
+      const more = getThemeProps({ theme, name, props: other });
+
+      // Provide the theme to the wrapped component.
+      // So we don't have to use the `withTheme()` Higher-order Component.
+      if (withTheme) {
+        more.theme = theme;
+      }
+
+      return (
+        <Component
+          ref={innerRef}
+          classes={getClasses({
+            classes,
+            theme,
+            stylesCreator,
+            cacheClasses: this.cacheClasses,
+            stylesOptions,
+            Component,
+          })}
+          {...more}
+        />
+      );
+    }
+  }
+
+  WithStylesInner.propTypes = {
+    /**
+     * Override or extend the styles applied to the component.
+     */
+    classes: PropTypes.object,
+    stylesOptions: PropTypes.object.isRequired,
+    theme: PropTypes.object,
+  };
+
+  const WithStyles = React.forwardRef((props, ref) => (
+    <StylesContext.Consumer>
+      {stylesOptions1 => {
+        const stylesOptions = {
+          ...stylesOptions1,
+          ...stylesOptions2,
+        };
+
+        return listenToTheme ? (
+          <ThemeContext.Consumer>
+            {theme => (
+              <WithStylesInner stylesOptions={stylesOptions} ref={ref} theme={theme} {...props} />
+            )}
+          </ThemeContext.Consumer>
+        ) : (
+          <WithStylesInner stylesOptions={stylesOptions} ref={ref} theme={noopTheme} {...props} />
+        );
+      }}
+    </StylesContext.Consumer>
+  ));
+
+  WithStyles.propTypes = {
+    /**
+     * Override or extend the styles applied to the component.
+     */
+    classes: PropTypes.object,
+    /**
+     * Use that property to pass a ref callback to the decorated component.
+     */
+    innerRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
```

Comment:
Can we get rid of this with ref forwarding?

Reply:
Not yet, there is a an issue when using forwardRef.

Reply:
@oliviertassinari What is the issue and why was `forwardRef` kept in this PR when there is an issue?